### ISSUE TITLE
Be able to target child elements on hover, with custom 'state'

### ISSUE THREAD

**Issue by wplit** [soflyy/oxygen-bugs-and-features#214]
**Describe the feature you'd like to see included in Oxygen.**
To be able to not just add :hover, :focus, :before, etc.. but to include more complicated targeting such as...

```
:hover .child-element
:hover + .next-element
:hover > .direct-child
::after:hover
```

<img width=500 src="https://i.ibb.co/g97gWXY/child.png">

Currently we can actually type this into Oxygen and it works. Until we save and view on the front end, then the `:hover .child-element` gets changed to `:hoverchild-element` and isn't actually applied. Clearly the states settings wasn't made for this, but it would open up the opportunity for lots of cool interactions just by using the default Oxygen controls.

**What are the use cases for this feature?**
Hover effects often include child elements to make them feel more natural/dynamic, otherwise they are very limited. 

A common use case would be changing the opacity of a featured image when the user hovers over the featured post. Not just when the user hovers over the image, but actually when they hover over the post.

I want to include in design sets, but custom CSS and selectors don't get carried over to other sites. So, apart from adding code blocks in to make up for the lacking CSS, it limits what we can create.

Of course we can already do in Oxygen, I'm just asking for it to be allowed in the 'state' options or something similar if this is awkward.

Mainly so simple hover effects can be part of design kits that are shared without having to add code blocks in, as it will be hard for users to customise if they are struggling finding where the CSS is being applied.

**Comment by louisreingold** [comment]
Assuming we do a dedicated UI for this instead of just writing it into the :state dropdown... does anyone have ideas for how that might look and work?

However, the idea of including this in the :state dropdown is a very good one.

Yeah, of course, it's a total hack, but this is the exact type of thing that a hacky solution like this is perfect for.

It's simple, it probably requires almost no code changes in Oxygen, and anyone who is going to be targeting child elements when the parent has a specific state already knows how to write a CSS selector.

**Comment by wplit** [comment]
If you were thinking of something more generic and flexible. In an ideal world for me, we'd be able to add anything we like to the end of selectors based on just the parent class, then we can include styles for unknown child elements and an unknown number of elements.

 **(Doesn't quite work under the 'state' dropdown though if it were this generic.)**

<img src="https://i.ibb.co/LNLD3JN/state.png" width=400>

The main use case I'm talking about is only for when we are building design sets meant for other users to use. I doubt many people will use.
 
For eg, if I build a specific layout of elements, but the user wants to use a different number of those elements, I don't want the design to break on them. I want to be able to build in such a way that is flexible enough so they can comfortably use it & customise without fear. eg Change the number of testimonials from 4 to 3 without it looking odd and having to change the styling every time. 

Or sometimes we want the last item in a section to have no bottom margin regardless of what element that is. (so the distance from the bottom of the section is just the padding and is the same as the top and doesn't look different)  So we need to add > *:last-child. All easy to do on our own projects where we have stylesheets and know the elements we're using. But more difficult to include in design sets without putting code blocks everywhere.

**Comment by wplit** [comment]
Could just be a second option built in that works pretty much the same as the state dropdown but allows for all this stuff.

<img width=400 src="https://i.ibb.co/sFdLnck/child.png">

**Comment by wplit** [comment]
Of course there's always doing it the way Webflow does.. using the parent element as a 'trigger' to add effects on the child elements. [https://www.youtube.com/watch?v=8g8MMCbfWTA](url)

**Comment by ivanjuras** [comment]
I thought we could already do this??? If not, please implement this. +1 :)

**Comment by wplit** [comment]
me too, it's because it does work in the builder. The problem is when it outputs the CSS on the front, it doesn't currently allow it.

**Comment by wplit** [comment]
For now, i've worked out a way of being able to change child elements on parent:hover without resorting to using code blocks after each element.

I can use CSS variables. It means not being able to use the visual controls for certain parts, still writing out the CSS, but it solves at least one problem, which was having to use code blocks to store CSS after every element. I want all the styles to be contained in that one element, to keep it cleaner.

Basically I can use variables on the child elements for things that are going to change, so things like opacity, transform, but then set those variables on the parent. So I can then go into the parent hover state, change the values of the variables and it ensures when the parent is hovered, all child elements using those variables will be effected.

**Comment by wplit** [comment]
But in regards to building this in properly so everybody can easily create child hover effects without resorting to writing CSS... I can image something like...

User selects parent element.
Clicks on states > hover
Then moves mouse over a child element.
Holds shift + clicks child element.
Selector changes to .parent:hover .child
User can now use visual editor to add styles to the child on parent hover.

**Comment by guyinpv** [comment]
I ran into this issue today and would like to see it built.
To do so without limitations would be very difficult I think, because this is no longer about a "state", it's about an entire CSS selector chain.
We're not only dealing with direct children, but possibly many grandchildren down. What if there are two hover states in the chain? `.menu li:hover > a > ul:hover` whatever.

So we're not just constructing a state of one element, but an entire CSS selector.
The state of an anchor link is not the same thing as the state of some parent wrapper somewhere. These are now two different elements with two different states. The parent wrapper might not DO anything with it's state, but it still has to be processed that such a state exists, in order for a child somewhere to make use of it.

So I have an alternate workflow. It's a kind of way of "enabling" a state to make it usable for child elements to latch on to.

For example, say you have a DOM structure like `div > ul > li > a > ul > li > a > span`

And now, hypothetically, this is a contrived example, you want the span to have an effect when they hover over the first anchor link. These are 3 elements apart. So you need something like `a:hover > ul  .span`
What I suggest is that you simply go find the parent selector of the first anchor, and then somehow "enable" the hover state. You don't have to change anything or set any styles, but just mark it that somewhere at some point you will make use of hover on this element.
Once you've enabled the hover state, you then drill down in the DOM and select the span. Now when you go to the State menu UI, it should show the "enabled" parent states you've selected to be available.

What Oxygen would do, then, is create a kind of suggested CSS selector using the parent state as the start. But also let us change the suggested selector if it is wrong somehow.

So once I've clicked on the anchor and enabled hover. I then click on the span and click on State button. In the menu would show something like "#that-specific-anchor:hover". If I were to click on this option, it would provide me a customized selector like "#that-specific-anchor:hover #this-specifc-span".

Now I might not want to use the specific IDs of those two elements. So this is where it would make sense to let me adjust it. Perhaps if the element also has a class, Oxygen would let me choose between the generic class, or the specific ID. So then it would change to ".generic-class:hover #specific-span".

And if I didn't want the specific span, and a class is available, I could switch to ".generic-class:hover .generic-span".
But what if I don't want a class? I should also be able to do this with just the element, such as ".generic-class:hover span".

Because there are so many ways to construct the new state selector, it has to be editable somehow. Maybe I even want to do "#that-specific-anchor > ul > li a > span".
Now that introduces all kinds of extra elements and child selector specificity. There is no easy way to construct this except to let me have edit access to a suggested selector.

Now whenever I select one of these spans, in the State menu it would show my customized selector. These selectors could also be recorded in the normal "Selectors" area of the Oxygen settings. Maybe a auto-generated category for "states" or something.

I think the above could work because it still makes use of the existing State menu, and it only requires two UI changes. One is a button or toggle or something to "enable" the state for use in child elements. This would operate separately from whether particular styles are set or not, and would not be deleted if the user clicks the X to remove state styles.
If the user later "disables" state, this would not change any in-use styles where that parent state was already selected and used. The only purpose for enabling a state is so that it shows up in the State menu of child elements. No other reason.

The 2nd UI change is when these enabled states show up in the list of States for child elements. It can look just like any other state choice, except there would need to be some method of adjusting and editing the selector. If the user customizes the selector such that it no longer even targets the current element, Oxygen could throw a warning or something that the selector is invalid for the current element.

One gotcha in this technique would be if the user enables state for multiple sub elements. That would make something like "div:hover > ul a:hover > span" possible. If hover state is enabled on both the div and on the anchor links. This could get awkward for Oxygen to handle in the State selector box, when it sees multiple parents have state enabled.

Anyway, that's my contribution to the problem!

**Comment by alexanderkladov** [comment]
Just wanted to pitch in, since I think that instead of [this suggestion](https://github.com/soflyy/oxygen-bugs-and-features/issues/214#issuecomment-480145069) perhaps it would make more sense to allow users to select a parent CSS selector like this:

<img src="https://i.ibb.co/WzcGYV5/Oxygen-Builder-Parent-CSS-Selector-feature-UI-suggestion.png" alt="Oxygen Builder Parent CSS Selector feature UI suggestion">

Why? Because it is more logical to apply the styling to the element being changed (i.e. `.child`), rather than it's parent, which just acts as a trigger (i.e. on `.parent:hover` we want to change `.child`).

So, for example, if we have the following layout:

```html
<div class="parent">
  <div class="child"></div>
</div>
```

Instead of applying the styles to `.parent` and adding `:hover .child` selector to it in order to target it's child on `:hover`, it would make much more sense to be styling the `.child` instead and add a parent CSS selector, if necessary to create unique effects on parent interaction. That way it's the `.child` that's going to have custom styling tied to it. In [this suggestion](https://github.com/soflyy/oxygen-bugs-and-features/issues/214#issuecomment-480145069) the styling would be tied to the parent and would be hard to find (until we get some sort of "view all page styles" feature from Oxygen).

**Comment by swinggraphics** [comment]
If I read the suggestions correctly, we would also be able to just use a class name as a "state", such as `.class-added-by-js`. This would allow us to style an element when JavaScript changes its class. +1

**Comment by guyinpv** [comment]
Perhaps we're over-complicating this.

At its core, all we're trying to do is latch the style editor onto a full CSS selector that may included parent states.

The only way to do this now is to use the stylesheets editor, or a custom code block with normal CSS. Perhaps the Selectors tool.

What we really need is the normal "Selectors" feature being expanded so that it can be used/edited from the sidebar style editor sidebar on the fly.
In other words, I can open the Selectors thing and add "ul:hover a span" as a selector, but I can't add this selector at the time I have the span selected in the editing window, nor toggle its state as I design/dev the page.

Imagine instead that we click on the span, and then from the sidebar, instead of only being able to add a state, or add a new class, we get to add a completely custom selector chain.

The only real gotcha is that if we use fancy features in the custom selector, like nth-child and :hover and so forth, it could make it difficult to toggle within the editing area. Maybe Oxygen could simply detect the use of a state within the selector chain and allow you to toggle that state on and off while editing? 

I don't have any clever ways of visually showing this right now. I guess I'm only talking about on-the-fly creation of custom selectors, combined with some intelligence to detect states and allow them on and off during design/dev time.

It's also possible we don't even need the intelligent toggle. Consider `ul:hover a span`, the only reason for this selector to exist is to style the hover state. Without :hover, all the elements would simply be their normal selves. I don't have to also style "ul a span" because those styles are set using all the normal selection UX. Regardless, the intelligence would be a convenience feature, so when I have the span selected in the editing interface, Oxygen could toggle that parent hover state ON while I still have my span selected. This only works if my custom state CSS chain is set up logically as a "state" in the first place. Not simply a plain old custom selector.

**Comment by swinggraphics** [comment]
I love the Selector Detector. I wish it were available for everything, all the time, as just another way to create a Selector. Then, any given element you select for editing should have a pane where it shows the custom Selectors that apply to it, making them much more convenient. Of course, all elements should be represented in Structure, selectable and editable, like header left/center/right, section containers, and post list divs, etc. The State selectors should be allowed to be anything. When you select one to apply for editing, the editor would take the rules that are set for the last item in the chain and apply them with `!important` for the preview. This works great _in theory_, of course. 😜

**Comment by schietegal** [comment]
That's one of the little things that make a huge difference IMO!

**Comment by Spellhammer** [comment]
All of the valid selectors indicated in the original request are now possible to use in Oxygen 4.2 and later.

Note that instead of :after:hover, you should use :hover:after to effect the :after pseudo-element when the parent is hovered.

I might be wrong but I don't think :after:hover is valid in CSS.
